### ISSUE TITLE
[JENKINS-44598] git workspace should be reset before build PR merged with base branch

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/github-branch-source-plugin#1085]
<p>The plugin attempts to merge the PR and the base branch before build. However, if git-merge introduces a conflict, the unmerged files will be left in the PR's git workspace, which causes git-merge in the following builds always fail, because the conflict is not solved.</p>

<p>Here's the log snippets:</p>

<p>For build #3, which introduces the merge conflict</p>

<pre>
> git fetch --tags --progress git@​github.mycompany.com:SERVER_REPO.git +refs/pull<span class="code-comment">//head:refs/remotes/origin/pr/
</span> Merging dev_6_2 commit fc999a1dae8861c1e5d3c21c8deb81f1e7bdf3a8 into PR head commit 5042e295186f8fe4bbb61ac97d97d69d1932912c
 > git config core.sparsecheckout # timeout=10
 > git checkout -f 5042e295186f8fe4bbb61ac97d97d69d1932912c
 > git merge fc999a1dae8861c1e5d3c21c8deb81f1e7bdf3a8 # timeout=10
 > git config core.sparsecheckout # timeout=10
 > git checkout -f 5042e295186f8fe4bbb61ac97d97d69d1932912c
GitHub has been notified of <span class="code-keyword">this</span> commit’s build result
hudson.plugins.git.GitException: Command <span class="code-quote">"git merge fc999a1dae8861c1e5d3c21c8deb81f1e7bdf3a8"</span> returned status code 1:
 stdout: Auto-merging docker-environment/ansible/group_vars/all.yml
 CONFLICT (add/add): Merge conflict in docker-environment/ansible/group_vars/all.yml
 Automatic merge failed; fix conflicts and then commit the result.
stderr: 
 at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:1877)
 at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:1845)
 at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:1841)
 at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommand(CliGitAPIImpl.java:1486)
 at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$3.execute(CliGitAPIImpl.java:621)
 at org.jenkinsci.plugins.github_branch_source.GitHubSCMSource$MergeWith.decorateRevisionToBuild(GitHubSCMSource.java:1003)
 at hudson.plugins.git.GitSCM.determineRevisionToBuild(GitSCM.java:1015)
 at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1108)
 at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:109)
 at org.jenkinsci.plugins.workflow.cps.CpsScmFlowDefinition.create(CpsScmFlowDefinition.java:130)
 at org.jenkinsci.plugins.workflow.multibranch.SCMBinder.create(SCMBinder.java:107)
 at org.jenkinsci.plugins.workflow.job.WorkflowRun.run(WorkflowRun.java:234)
 at hudson.model.ResourceController.execute(ResourceController.java:98)
 at hudson.model.Executor.run(Executor.java:405)
 Finished: FAILURE
</pre>

<p> For the following build #4, #5 and so on, the merge always fails</p>

<pre>
> git fetch --tags --progress git@​github.mycompany.com:SERVER_REPO.git +refs/pull<span class="code-comment">//head:refs/remotes/origin/pr/
</span> Merging dev_6_2 commit fc999a1dae8861c1e5d3c21c8deb81f1e7bdf3a8 into PR head commit 32fe21329c3f0349347d029917ba29a39fc0364a
 > git config core.sparsecheckout # timeout=10
 > git checkout -f 32fe21329c3f0349347d029917ba29a39fc0364a
 > git merge fc999a1dae8861c1e5d3c21c8deb81f1e7bdf3a8 # timeout=10
 > git config core.sparsecheckout # timeout=10
 > git checkout -f 32fe21329c3f0349347d029917ba29a39fc0364a
GitHub has been notified of <span class="code-keyword">this</span> commit’s build result
hudson.plugins.git.GitException: Command <span class="code-quote">"git merge fc999a1dae8861c1e5d3c21c8deb81f1e7bdf3a8"</span> returned status code 128:
 stdout: 
 stderr: error: <span class="code-quote">'merge'</span> is not possible because you have unmerged files.
 hint: Fix them up in the work tree,
 hint: and then use <span class="code-quote">'git add/rm <file>'</span> as
 hint: appropriate to mark resolution and make a commit,
 hint: or use <span class="code-quote">'git commit -a'</span>.
 fatal: Exiting because of an unresolved conflict.
at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:1877)
 at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:1845)
 at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:1841)
 at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommand(CliGitAPIImpl.java:1486)
 at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$3.execute(CliGitAPIImpl.java:621)
 at org.jenkinsci.plugins.github_branch_source.GitHubSCMSource$MergeWith.decorateRevisionToBuild(GitHubSCMSource.java:1003)
 at hudson.plugins.git.GitSCM.determineRevisionToBuild(GitSCM.java:1015)
 at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1108)
 at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:109)
 at org.jenkinsci.plugins.workflow.cps.CpsScmFlowDefinition.create(CpsScmFlowDefinition.java:130)
 at org.jenkinsci.plugins.workflow.multibranch.SCMBinder.create(SCMBinder.java:107)
 at org.jenkinsci.plugins.workflow.job.WorkflowRun.run(WorkflowRun.java:234)
 at hudson.model.ResourceController.execute(ResourceController.java:98)
 at hudson.model.Executor.run(Executor.java:405)
 Finished: FAILURE
</pre>

<p>  </p>

<p>Then this build will be stuck in this situation unless we manually issue 'git reset HEAD --hard' in that git workspace.</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=asouchang">asouchang</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44598" target="_blank">git workspace should be reset before build PR merged with base branch</a></i></summary>
<i><ul>
<li><b>status</b>: Open
<li><b>priority</b>: Major
<li><b>component(s)</b>: github-branch-source-plugin
<li><b>resolution</b>: Unresolved
<li><b>votes</b>: 9
<li><b>watchers</b>: 16
<li><b>imported</b>: 2025-12-02
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>The plugin attempts to merge the PR and the base branch before build. However, if git-merge introduces a conflict, the unmerged files will be left in the PR's git workspace, which causes git-merge in the following builds always fail, because the conflict is not solved.</p>

<p>Here's the log snippets:</p>

<p>For build #3, which introduces the merge conflict</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">&gt; git fetch --tags --progress git@github.mycompany.com:SERVER_REPO.git +refs/pull<span class="code-comment">//head:refs/remotes/origin/pr/
</span> Merging dev_6_2 commit fc999a1dae8861c1e5d3c21c8deb81f1e7bdf3a8 into PR head commit 5042e295186f8fe4bbb61ac97d97d69d1932912c
 &gt; git config core.sparsecheckout # timeout=10
 &gt; git checkout -f 5042e295186f8fe4bbb61ac97d97d69d1932912c
 &gt; git merge fc999a1dae8861c1e5d3c21c8deb81f1e7bdf3a8 # timeout=10
 &gt; git config core.sparsecheckout # timeout=10
 &gt; git checkout -f 5042e295186f8fe4bbb61ac97d97d69d1932912c
GitHub has been notified of <span class="code-keyword">this</span> commit’s build result
hudson.plugins.git.GitException: Command <span class="code-quote">"git merge fc999a1dae8861c1e5d3c21c8deb81f1e7bdf3a8"</span> returned status code 1:
 stdout: Auto-merging docker-environment/ansible/group_vars/all.yml
 CONFLICT (add/add): Merge conflict in docker-environment/ansible/group_vars/all.yml
 Automatic merge failed; fix conflicts and then commit the result.
stderr: 
 at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:1877)
 at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:1845)
 at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:1841)
 at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommand(CliGitAPIImpl.java:1486)
 at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$3.execute(CliGitAPIImpl.java:621)
 at org.jenkinsci.plugins.github_branch_source.GitHubSCMSource$MergeWith.decorateRevisionToBuild(GitHubSCMSource.java:1003)
 at hudson.plugins.git.GitSCM.determineRevisionToBuild(GitSCM.java:1015)
 at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1108)
 at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:109)
 at org.jenkinsci.plugins.workflow.cps.CpsScmFlowDefinition.create(CpsScmFlowDefinition.java:130)
 at org.jenkinsci.plugins.workflow.multibranch.SCMBinder.create(SCMBinder.java:107)
 at org.jenkinsci.plugins.workflow.job.WorkflowRun.run(WorkflowRun.java:234)
 at hudson.model.ResourceController.execute(ResourceController.java:98)
 at hudson.model.Executor.run(Executor.java:405)
 Finished: FAILURE
</pre>
</div></div>

<p> For the following build #4, #5 and so on, the merge always fails</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">&gt; git fetch --tags --progress git@github.mycompany.com:SERVER_REPO.git +refs/pull<span class="code-comment">//head:refs/remotes/origin/pr/
</span> Merging dev_6_2 commit fc999a1dae8861c1e5d3c21c8deb81f1e7bdf3a8 into PR head commit 32fe21329c3f0349347d029917ba29a39fc0364a
 &gt; git config core.sparsecheckout # timeout=10
 &gt; git checkout -f 32fe21329c3f0349347d029917ba29a39fc0364a
 &gt; git merge fc999a1dae8861c1e5d3c21c8deb81f1e7bdf3a8 # timeout=10
 &gt; git config core.sparsecheckout # timeout=10
 &gt; git checkout -f 32fe21329c3f0349347d029917ba29a39fc0364a
GitHub has been notified of <span class="code-keyword">this</span> commit’s build result
hudson.plugins.git.GitException: Command <span class="code-quote">"git merge fc999a1dae8861c1e5d3c21c8deb81f1e7bdf3a8"</span> returned status code 128:
 stdout: 
 stderr: error: <span class="code-quote">'merge'</span> is not possible because you have unmerged files.
 hint: Fix them up in the work tree,
 hint: and then use <span class="code-quote">'git add/rm &lt;file&gt;'</span> as
 hint: appropriate to mark resolution and make a commit,
 hint: or use <span class="code-quote">'git commit -a'</span>.
 fatal: Exiting because of an unresolved conflict.
at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:1877)
 at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:1845)
 at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:1841)
 at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommand(CliGitAPIImpl.java:1486)
 at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$3.execute(CliGitAPIImpl.java:621)
 at org.jenkinsci.plugins.github_branch_source.GitHubSCMSource$MergeWith.decorateRevisionToBuild(GitHubSCMSource.java:1003)
 at hudson.plugins.git.GitSCM.determineRevisionToBuild(GitSCM.java:1015)
 at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1108)
 at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:109)
 at org.jenkinsci.plugins.workflow.cps.CpsScmFlowDefinition.create(CpsScmFlowDefinition.java:130)
 at org.jenkinsci.plugins.workflow.multibranch.SCMBinder.create(SCMBinder.java:107)
 at org.jenkinsci.plugins.workflow.job.WorkflowRun.run(WorkflowRun.java:234)
 at hudson.model.ResourceController.execute(ResourceController.java:98)
 at hudson.model.Executor.run(Executor.java:405)
 Finished: FAILURE
</pre>
</div></div>

<p>  </p>

<p>Then this build will be stuck in this situation unless we manually issue 'git reset HEAD --hard' in that git workspace.</p></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
OS: CentOS 7.3.1611<br/>
Jenkins: 2.46.2<br/>
GitHub Branch Source Plugin: 2.0.6<br/>
git version: 1.8.3.1
```
</details>
<details><summary><i>3 attachments</i></summary>

- [Screen Shot 2017-11-07 at 11.28.43.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/40242/Screen%20Shot%202017-11-07%20at%2011.28.43.png)
  > ![Screen Shot 2017-11-07 at 11.28.43.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/40242/Screen%20Shot%202017-11-07%20at%2011.28.43.png)
- [image-2023-06-20-09-50-58-110.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/60815/image-2023-06-20-09-50-58-110.png)
  > ![image-2023-06-20-09-50-58-110.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/60815/image-2023-06-20-09-50-58-110.png)
- [image-2023-07-13-08-40-45-282.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/60816/image-2023-07-13-08-40-45-282.png)
  > ![image-2023-07-13-08-40-45-282.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/60816/image-2023-07-13-08-40-45-282.png)
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/medianick.png" title="medianick's avatar" /> medianick:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44598?focusedId=306363&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-306363">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I just observed the same thing today, running GitHub Branch Source 2.0.7, git 2.13.2, and Jenkins 2.60.1 (LTS). The only solution was to track down the appropriate multibranch pipeline directory on the master node (which doesn't otherwise do builds, but where the merge conflict and the unmerged changes occurred), and run 'git reset HEAD --hard' there. </p>

<p>Even if it didn't make sense to do resets automatically in all cases prior to attempting to merge, it would be nice to have the ability to add behaviors to the GitHub SCM for Multibranch Pipeline jobs as is possible with Freestyle jobs.</p>
</pre>
</details>

</ul>
</details>

<p>I just observed the same thing today, running GitHub Branch Source 2.0.7, git 2.13.2, and Jenkins 2.60.1 (LTS). The only solution was to track down the appropriate multibranch pipeline directory on the master node (which doesn't otherwise do builds, but where the merge conflict and the unmerged changes occurred), and run 'git reset HEAD --hard' there. </p>

<p>Even if it didn't make sense to do resets automatically in all cases prior to attempting to merge, it would be nice to have the ability to add behaviors to the GitHub SCM for Multibranch Pipeline jobs as is possible with Freestyle jobs.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/stephenconnolly.png" title="stephenconnolly's avatar" /> stephenconnolly:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44598?focusedId=308437&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-308437">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=medianick" class="user-hover" rel="medianick">medianick</a> with <a href="https://issues.jenkins.io/browse/JENKINS-43507" title="Allow SCMSource and SCMNavigator subtypes to share common traits" class="issue-link" data-issue-key="JENKINS-43507"><del>JENKINS-43507</del></a> you can now add the Clean Before Checkout behaviour, so that should at least provide a workaround.</p>

<p>More generally, we probably should apply this behaviour fix after <a href="https://issues.jenkins.io/browse/JENKINS-45771" title="Disable shallow clone when we know a merge will take place" class="issue-link" data-issue-key="JENKINS-45771"><del>JENKINS-45771</del></a> has been merged in order to ensure the fix is picked up by all git-based scm source implementations</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=medianick" class="user-hover" rel="medianick">medianick</a> with <a href="https://issues.jenkins.io/browse/JENKINS-43507" title="Allow SCMSource and SCMNavigator subtypes to share common traits" class="issue-link" data-issue-key="JENKINS-43507"><del>JENKINS-43507</del></a> you can now add the Clean Before Checkout behaviour, so that should at least provide a workaround.</p>

<p>More generally, we probably should apply this behaviour fix after <a href="https://issues.jenkins.io/browse/JENKINS-45771" title="Disable shallow clone when we know a merge will take place" class="issue-link" data-issue-key="JENKINS-45771"><del>JENKINS-45771</del></a> has been merged in order to ensure the fix is picked up by all git-based scm source implementations</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>acejam:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44598?focusedId=314060&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-314060">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I'm also now seeing this exact issue too. Any ideas around a fix? If someone can point me in the right direction I can attempt one.</p>
</pre>
</details>

</ul>
</details>

<p>I'm also now seeing this exact issue too. Any ideas around a fix? If someone can point me in the right direction I can attempt one.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/stephenconnolly.png" title="stephenconnolly's avatar" /> stephenconnolly:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44598?focusedId=314068&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-314068">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p> add the Clean Before Checkout behaviour</p>
</pre>
</details>

</ul>
</details>

<p> add the Clean Before Checkout behaviour</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/trejkaz.jpg" title="trejkaz's avatar" /> trejkaz:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44598?focusedId=318985&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-318985">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Clean Before Checkout didn't help for us... maybe cleaning isn't cleaning it enough?</p>

<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>Cleaning workspace
 &gt; git rev-parse --verify HEAD # timeout=10
Resetting working tree
 &gt; git reset --hard # timeout=10
 &gt; git clean -fdx # timeout=10
Pruning obsolete local branches
Fetching upstream changes from git@github.eng.acme.com:acme/core.git
 &gt; git --version # timeout=10
using GIT_SSH to set credentials GitHub BuildSlaves Account (Pipeline-Github)
 &gt; git fetch --tags --progress git@github.eng.acme.com:acme/core.git +refs/pull/145/head:refs/remotes/origin/PR-145 +refs/heads/master:refs/remotes/origin/master --prune # timeout=40
Merging remotes/origin/master commit fb359ab126eef7794d4118f45c9057695d422162 into PR head commit dc24fa110e813e0c43a2230db6d319f5ea7c11b9
 &gt; git config core.sparsecheckout # timeout=10
 &gt; git checkout -f dc24fa110e813e0c43a2230db6d319f5ea7c11b9 # timeout=40
 &gt; git merge fb359ab126eef7794d4118f45c9057695d422162 # timeout=10
 &gt; git config core.sparsecheckout # timeout=10
 &gt; git checkout -f dc24fa110e813e0c43a2230db6d319f5ea7c11b9 # timeout=40
Command "git merge fb359ab126eef7794d4118f45c9057695d422162" returned status code 1:
stdout: Auto-merging processing/src/test/java/com/acme/export/print/TestAsposeVisioPdfExporter.java
Auto-merging investigator/src/resources/com/acme/investigator/charts/acme_metadatarequest.js
CONFLICT (add/add): Merge conflict in investigator/src/resources/com/acme/investigator/charts/acme_metadatarequest.js
Auto-merging investigator/src/resources/com/acme/investigator/charts/main.html
Auto-merging investigator/src/resources/com/acme/investigator/charts/datamanager.js
Auto-merging investigator/src/java/com/acme/investigator/charts/ChartsController.java
CONFLICT (content): Merge conflict in investigator/src/java/com/acme/investigator/charts/ChartsController.java
Removing gui-util/src/resources/com/acme/browser/resources/test.html
Auto-merging buildSrc/hooks/pre-commit
CONFLICT (add/add): Merge conflict in buildSrc/hooks/pre-commit
Auto-merging build.gradle
CONFLICT (content): Merge conflict in build.gradle
Auto-merging Jenkinsfile
CONFLICT (content): Merge conflict in Jenkinsfile
Automatic merge failed; fix conflicts and then commit the result.
</pre>
</div></div>

<p>At this rate it feels like the only workaround is to blow away the checkout directory before starting the build, like what we had to do with svn due to similar sorts of issues...</p>
</pre>
</details>

</ul>
</details>

<p>Clean Before Checkout didn't help for us... maybe cleaning isn't cleaning it enough?</p>



```
Cleaning workspace
 > git rev-parse --verify HEAD # timeout=10
Resetting working tree
 > git reset --hard # timeout=10
 > git clean -fdx # timeout=10
Pruning obsolete local branches
Fetching upstream changes from git@​github.eng.acme.com:acme/core.git
 > git --version # timeout=10
using GIT_SSH to set credentials GitHub BuildSlaves Account (Pipeline-Github)
 > git fetch --tags --progress git@​github.eng.acme.com:acme/core.git +refs/pull/145/head:refs/remotes/origin/PR-145 +refs/heads/master:refs/remotes/origin/master --prune # timeout=40
Merging remotes/origin/master commit fb359ab126eef7794d4118f45c9057695d422162 into PR head commit dc24fa110e813e0c43a2230db6d319f5ea7c11b9
 > git config core.sparsecheckout # timeout=10
 > git checkout -f dc24fa110e813e0c43a2230db6d319f5ea7c11b9 # timeout=40
 > git merge fb359ab126eef7794d4118f45c9057695d422162 # timeout=10
 > git config core.sparsecheckout # timeout=10
 > git checkout -f dc24fa110e813e0c43a2230db6d319f5ea7c11b9 # timeout=40
Command "git merge fb359ab126eef7794d4118f45c9057695d422162" returned status code 1:
stdout: Auto-merging processing/src/test/java/com/acme/export/print/TestAsposeVisioPdfExporter.java
Auto-merging investigator/src/resources/com/acme/investigator/charts/acme_metadatarequest.js
CONFLICT (add/add): Merge conflict in investigator/src/resources/com/acme/investigator/charts/acme_metadatarequest.js
Auto-merging investigator/src/resources/com/acme/investigator/charts/main.html
Auto-merging investigator/src/resources/com/acme/investigator/charts/datamanager.js
Auto-merging investigator/src/java/com/acme/investigator/charts/ChartsController.java
CONFLICT (content): Merge conflict in investigator/src/java/com/acme/investigator/charts/ChartsController.java
Removing gui-util/src/resources/com/acme/browser/resources/test.html
Auto-merging buildSrc/hooks/pre-commit
CONFLICT (add/add): Merge conflict in buildSrc/hooks/pre-commit
Auto-merging build.gradle
CONFLICT (content): Merge conflict in build.gradle
Auto-merging Jenkinsfile
CONFLICT (content): Merge conflict in Jenkinsfile
Automatic merge failed; fix conflicts and then commit the result.

```

<p>At this rate it feels like the only workaround is to blow away the checkout directory before starting the build, like what we had to do with svn due to similar sorts of issues...</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/stephenconnolly.png" title="stephenconnolly's avatar" /> stephenconnolly:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44598?focusedId=318991&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-318991">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=trejkaz" class="user-hover" rel="trejkaz">trejkaz</a> is the PR mergable without conflicts? If it is not mergable then the merge should fail. What does GitHub say about the merge status?</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=trejkaz" class="user-hover" rel="trejkaz">trejkaz</a> is the PR mergable without conflicts? If it is not mergable then the merge should fail. What does GitHub say about the merge status?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/trejkaz.jpg" title="trejkaz's avatar" /> trejkaz:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44598?focusedId=318992&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-318992">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>GitHub's merge status isn't really useful - it just says "Required statuses must pass before merging" because Jenkins failed the build.</p>

<p>I imagine that if Jenkins passed, it would just let me merge, because the PR only changes two files, neither of which are the ones the command is complaining about.</p>

<p>Additionally, the output seems a bit suspicious:</p>

<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre> &gt; git checkout -f dc24fa110e813e0c43a2230db6d319f5ea7c11b9 # timeout=40
Command "git merge fb359ab126eef7794d4118f45c9057695d422162" returned status code 1:
</pre>
</div></div>

<p>Shouldn't <tt>git merge</tt> be the last command output before that failure? </p>

<p>(Also, why is it running <tt>git checkout</tt> twice?)</p>

<p>In an attempt to reproduce it locally, I checked the same sequence of commands locally, and they worked fine on my box, but I'm running macOS and the build machine is running Debian, so it isn't a realistic test.</p>
</pre>
</details>

</ul>
</details>

<p>GitHub's merge status isn't really useful - it just says "Required statuses must pass before merging" because Jenkins failed the build.</p>

<p>I imagine that if Jenkins passed, it would just let me merge, because the PR only changes two files, neither of which are the ones the command is complaining about.</p>

<p>Additionally, the output seems a bit suspicious:</p>



```
 > git checkout -f dc24fa110e813e0c43a2230db6d319f5ea7c11b9 # timeout=40
Command "git merge fb359ab126eef7794d4118f45c9057695d422162" returned status code 1:

```

<p>Shouldn't <tt>git merge</tt> be the last command output before that failure? </p>

<p>(Also, why is it running <tt>git checkout</tt> twice?)</p>

<p>In an attempt to reproduce it locally, I checked the same sequence of commands locally, and they worked fine on my box, but I'm running macOS and the build machine is running Debian, so it isn't a realistic test.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/stephenconnolly.png" title="stephenconnolly's avatar" /> stephenconnolly:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44598?focusedId=319028&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-319028">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=trejkaz" class="user-hover" rel="trejkaz">trejkaz</a> Aha!</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">git config core.sparsecheckout # timeout=10</pre>
</div></div>
<p>You cannot merge if you are doing a sparse checkout, because git doesn't have the history with which to do a real merge. </p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/40242/40242_Screen+Shot+2017-11-07+at+11.28.43.png" width="600" style="border: 0px solid black" /></span></p>

<p>Can you try turning off "Shallow clone" in "Advanced clone behaviours". If that fixes the issue then likely we need to configure the GitHub Branch Source to override shallow clone configuration for PR-merge SCMHeads</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=trejkaz" class="user-hover" rel="trejkaz">trejkaz</a> Aha!</p>

<pre>
git config core.sparsecheckout # timeout=10</pre>
<p>You cannot merge if you are doing a sparse checkout, because git doesn't have the history with which to do a real merge. </p>

<p><span class="image-wrap" style=""><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/40242/Screen%20Shot%202017-11-07%20at%2011.28.43.png" width="600" style="border: 0px solid black" /></span></p>

<p>Can you try turning off "Shallow clone" in "Advanced clone behaviours". If that fixes the issue then likely we need to configure the GitHub Branch Source to override shallow clone configuration for PR-merge SCMHeads</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/trejkaz.jpg" title="trejkaz's avatar" /> trejkaz:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44598?focusedId=319241&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-319241">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Looks like "Shallow clone" is already off. "Shallow clone depth" is 10... but I'm hoping it's ignored if the setting isn't even turned on. (Maybe that is a false hope, though?)</p>

<p>I tried setting it to blank to see if it changes anything. It ends up being set to 0... I can't get it back to an empty box like shown in your screenshot. :/</p>
</pre>
</details>

</ul>
</details>

<p>Looks like "Shallow clone" is already off. "Shallow clone depth" is 10... but I'm hoping it's ignored if the setting isn't even turned on. (Maybe that is a false hope, though?)</p>

<p>I tried setting it to blank to see if it changes anything. It ends up being set to 0... I can't get it back to an empty box like shown in your screenshot. :/</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/trejkaz.jpg" title="trejkaz's avatar" /> trejkaz:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44598?focusedId=319247&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-319247">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>OK, it's still running `git config core.sparsecheckout` even with both fields reset, so maybe `git config core.sparsecheckout` just means to turn the setting off? :/</p>
</pre>
</details>

</ul>
</details>

<p>OK, it's still running `git config core.sparsecheckout` even with both fields reset, so maybe `git config core.sparsecheckout` just means to turn the setting off? :/</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>drewish:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44598?focusedId=328523&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-328523">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>The clean before build worked for me. The sparse checkout option didn't seem to have any effect on this issue.</p>
</pre>
</details>

</ul>
</details>

<p>The clean before build worked for me. The sparse checkout option didn't seem to have any effect on this issue.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>rthirucs:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44598?focusedId=357046&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-357046">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I'm still facing the problem. The Jira issue is in OPEN state. Can this be resolved somehow?</p>
</pre>
</details>

</ul>
</details>

<p>I'm still facing the problem. The Jira issue is in OPEN state. Can this be resolved somehow?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>brianjmurrell:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44598?focusedId=360535&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-360535">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I too am hitting this and would appreciate a status update on whether anything is being done to resolve it.</p>

<p>Adding the <b>Clean before checkout</b> (from Jenkins Git plugin) isn't really an acceptable solution because I really only want the cleaning before the initial checkout on the master (just to resolve the initial checkout merge failure before the Pipeline is even run) and not before every checkout for every Pipeline build step.  This is because cleaning in some of the subsequent Pipeline steps leads to other errors such as:</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>10:21:35  Fetching changes from the remote Git repository
10:21:35  Cleaning workspace
10:21:35  ERROR: Error fetching remote repo 'origin'
10:21:35  hudson.plugins.git.GitException: Failed to fetch from https://github.com/daos-stack/daos.git
10:21:35  	at hudson.plugins.git.GitSCM.fetchFrom(GitSCM.java:904)
10:21:35  	at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1144)
10:21:35  	at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1175)
10:21:35  	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:120)
10:21:35  	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:90)
10:21:35  	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:77)
10:21:35  	at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
10:21:35  	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
10:21:35  	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
10:21:35  	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
10:21:35  	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
10:21:35  	at java.lang.Thread.run(Thread.java:748)
10:21:35  Caused by: hudson.plugins.git.GitException: Command "git clean -fdx" returned status code 128:
10:21:35  stdout: 
10:21:35  stderr: fatal: Not a git repository: install/include/boost/../../.git/modules/src/boost
10:21:35  
10:21:35  	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2318)
10:21:35  	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2248)
10:21:35  	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2244)
10:21:35  	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommand(CliGitAPIImpl.java:1777)
10:21:35  	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommand(CliGitAPIImpl.java:1789)
10:21:35  	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.clean(CliGitAPIImpl.java:901)
10:21:35  	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.clean(CliGitAPIImpl.java:912)
10:21:35  	at hudson.plugins.git.GitAPI.clean(GitAPI.java:311)
10:21:35  	at sun.reflect.GeneratedMethodAccessor94.invoke(Unknown Source)
10:21:35  	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
10:21:35  	at java.lang.reflect.Method.invoke(Method.java:498)
10:21:35  	at hudson.remoting.RemoteInvocationHandler$RPCRequest.perform(RemoteInvocationHandler.java:929)
10:21:35  	at hudson.remoting.RemoteInvocationHandler$RPCRequest.call(RemoteInvocationHandler.java:903)
10:21:35  	at hudson.remoting.RemoteInvocationHandler$RPCRequest.call(RemoteInvocationHandler.java:855)
10:21:35  	at hudson.remoting.UserRequest.perform(UserRequest.java:212)
10:21:35  	at hudson.remoting.UserRequest.perform(UserRequest.java:54)
10:21:35  	at hudson.remoting.Request$2.run(Request.java:369)
10:21:35  	at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:72)
10:21:35  	... 4 more
10:21:35  	Suppressed: hudson.remoting.Channel$CallSiteStackTrace: Remote call to wolf-33_docker_1
10:21:35  		at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1741)
10:21:35  		at hudson.remoting.UserRequest$ExceptionResponse.retrieve(UserRequest.java:357)
10:21:35  		at hudson.remoting.Channel.call(Channel.java:955)
10:21:35  		at hudson.remoting.RemoteInvocationHandler.invoke(RemoteInvocationHandler.java:283)
10:21:35  		at com.sun.proxy.$Proxy102.clean(Unknown Source)
10:21:35  		at org.jenkinsci.plugins.gitclient.RemoteGitImpl.clean(RemoteGitImpl.java:463)
10:21:35  		at hudson.plugins.git.extensions.impl.CleanBeforeCheckout.decorateFetchCommand(CleanBeforeCheckout.java:30)
10:21:35  		at hudson.plugins.git.GitSCM.fetchFrom(GitSCM.java:900)
10:21:35  		at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1144)
10:21:35  		at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1175)
10:21:35  		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:120)
10:21:35  		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:90)
10:21:35  		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:77)
10:21:35  		at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
10:21:35  		at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
10:21:35  		... 4 more
</pre>
</div></div>
<p>Yes, to be sure, the above is only a problem because of the content of the tree in the workspace, but it's only a problem because of this needed explicit <b>Clean before checkout</b> option that really should be the implied-and-always-executed behaviour when doing the initial merge checkout.  IOW, adding the explicit behaviour should not be necessary.</p>
</pre>
</details>

</ul>
</details>

<p>I too am hitting this and would appreciate a status update on whether anything is being done to resolve it.</p>

<p>Adding the <b>Clean before checkout</b> (from Jenkins Git plugin) isn't really an acceptable solution because I really only want the cleaning before the initial checkout on the master (just to resolve the initial checkout merge failure before the Pipeline is even run) and not before every checkout for every Pipeline build step.  This is because cleaning in some of the subsequent Pipeline steps leads to other errors such as:</p>


```
10:21:35  Fetching changes from the remote Git repository
10:21:35  Cleaning workspace
10:21:35  ERROR: Error fetching remote repo 'origin'
10:21:35  hudson.plugins.git.GitException: Failed to fetch from https://github.com/daos-stack/daos.git
10:21:35  	at hudson.plugins.git.GitSCM.fetchFrom(GitSCM.java:904)
10:21:35  	at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1144)
10:21:35  	at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1175)
10:21:35  	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:120)
10:21:35  	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:90)
10:21:35  	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:77)
10:21:35  	at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
10:21:35  	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
10:21:35  	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
10:21:35  	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
10:21:35  	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
10:21:35  	at java.lang.Thread.run(Thread.java:748)
10:21:35  Caused by: hudson.plugins.git.GitException: Command "git clean -fdx" returned status code 128:
10:21:35  stdout: 
10:21:35  stderr: fatal: Not a git repository: install/include/boost/../../.git/modules/src/boost
10:21:35  
10:21:35  	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2318)
10:21:35  	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2248)
10:21:35  	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2244)
10:21:35  	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommand(CliGitAPIImpl.java:1777)
10:21:35  	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommand(CliGitAPIImpl.java:1789)
10:21:35  	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.clean(CliGitAPIImpl.java:901)
10:21:35  	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.clean(CliGitAPIImpl.java:912)
10:21:35  	at hudson.plugins.git.GitAPI.clean(GitAPI.java:311)
10:21:35  	at sun.reflect.GeneratedMethodAccessor94.invoke(Unknown Source)
10:21:35  	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
10:21:35  	at java.lang.reflect.Method.invoke(Method.java:498)
10:21:35  	at hudson.remoting.RemoteInvocationHandler$RPCRequest.perform(RemoteInvocationHandler.java:929)
10:21:35  	at hudson.remoting.RemoteInvocationHandler$RPCRequest.call(RemoteInvocationHandler.java:903)
10:21:35  	at hudson.remoting.RemoteInvocationHandler$RPCRequest.call(RemoteInvocationHandler.java:855)
10:21:35  	at hudson.remoting.UserRequest.perform(UserRequest.java:212)
10:21:35  	at hudson.remoting.UserRequest.perform(UserRequest.java:54)
10:21:35  	at hudson.remoting.Request$2.run(Request.java:369)
10:21:35  	at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:72)
10:21:35  	... 4 more
10:21:35  	Suppressed: hudson.remoting.Channel$CallSiteStackTrace: Remote call to wolf-33_docker_1
10:21:35  		at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1741)
10:21:35  		at hudson.remoting.UserRequest$ExceptionResponse.retrieve(UserRequest.java:357)
10:21:35  		at hudson.remoting.Channel.call(Channel.java:955)
10:21:35  		at hudson.remoting.RemoteInvocationHandler.invoke(RemoteInvocationHandler.java:283)
10:21:35  		at com.sun.proxy.$Proxy102.clean(Unknown Source)
10:21:35  		at org.jenkinsci.plugins.gitclient.RemoteGitImpl.clean(RemoteGitImpl.java:463)
10:21:35  		at hudson.plugins.git.extensions.impl.CleanBeforeCheckout.decorateFetchCommand(CleanBeforeCheckout.java:30)
10:21:35  		at hudson.plugins.git.GitSCM.fetchFrom(GitSCM.java:900)
10:21:35  		at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1144)
10:21:35  		at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1175)
10:21:35  		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:120)
10:21:35  		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:90)
10:21:35  		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:77)
10:21:35  		at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
10:21:35  		at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
10:21:35  		... 4 more

```
<p>Yes, to be sure, the above is only a problem because of the content of the tree in the workspace, but it's only a problem because of this needed explicit <b>Clean before checkout</b> option that really should be the implied-and-always-executed behaviour when doing the initial merge checkout.  IOW, adding the explicit behaviour should not be necessary.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>bharathr0209:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44598?focusedId=404357&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-404357">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=brianjmurrell" class="user-hover" rel="brianjmurrell">brianjmurrell</a> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=rthirucs" class="user-hover" rel="rthirucs">rthirucs</a> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=trejkaz" class="user-hover" rel="trejkaz">trejkaz</a> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=asouchang" class="user-hover" rel="asouchang">asouchang</a> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=drewish" class="user-hover" rel="drewish">drewish</a>  Did somebody got this working?
 I get the below error: 
 <b>stderr: error: Your local changes to the following files would be overwritten by merge: Zipfile.zip</b></p>

<p>This started after I enabled <b>Git LFS pull after checkout</b> option in Multibranch Bitbucket pipeline,
 So i see this option doing a spare checkout as below
 <b>Enabling Git LFS pull</b>
 <b>&gt; git config core.sparsecheckout # timeout=10</b>
 And looks like as per the discussion above, with spare checkout, merge won't work?</p>

<p>Also, this happens on master node, even before my script execution starts.
 Clean before/after checkout, Wipeout repository &amp; force clone, none of them help.</p>

<p>The only workaround I have at the moment is to rebase the PR to master. This somehow solves the problem</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=brianjmurrell" class="user-hover" rel="brianjmurrell">brianjmurrell</a> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=rthirucs" class="user-hover" rel="rthirucs">rthirucs</a> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=trejkaz" class="user-hover" rel="trejkaz">trejkaz</a> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=asouchang" class="user-hover" rel="asouchang">asouchang</a> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=drewish" class="user-hover" rel="drewish">drewish</a>  Did somebody got this working?<br/>
 I get the below error: <br/>
 <b>stderr: error: Your local changes to the following files would be overwritten by merge: Zipfile.zip</b></p>

<p>This started after I enabled <b>Git LFS pull after checkout</b> option in Multibranch Bitbucket pipeline,<br/>
 So i see this option doing a spare checkout as below<br/>
 <b>Enabling Git LFS pull</b><br/>
 <b>> git config core.sparsecheckout # timeout=10</b><br/>
 And looks like as per the discussion above, with spare checkout, merge won't work?</p>

<p>Also, this happens on master node, even before my script execution starts.<br/>
 Clean before/after checkout, Wipeout repository & force clone, none of them help.</p>

<p>The only workaround I have at the moment is to rebase the PR to master. This somehow solves the problem</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>kon:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44598?focusedId=404362&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-404362">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Seeing "git config core.sparsecheckout" in the log does not mean that sparse checkout is enabled. git-client-plugin (<a href="https://github.com/jenkinsci/git-client-plugin/blob/cba18fb8a531fb19897f6455f12430b157eeb47b/src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java#L2815" class="external-link" target="_blank" rel="nofollow noopener">source</a>) runs that command just to ask Git whether sparse checkout is already enabled. Depending on how Git answers and whether you have configured "Sparse Checkout paths" in Jenkins, git-client-plugin can then run "git config core.sparsecheckout true" or "git config core.sparsecheckout false" to change the setting.</p>

<p>Also, shallow clone and sparse checkout are separate features. Shallow clone controls how much history Git clones from the remote repository to the local repository, and sparse checkout controls which files Git checks out from the local repository to the working tree.</p>
</pre>
</details>

</ul>
</details>

<p>Seeing "git config core.sparsecheckout" in the log does not mean that sparse checkout is enabled. git-client-plugin (<a href="https://github.com/jenkinsci/git-client-plugin/blob/cba18fb8a531fb19897f6455f12430b157eeb47b/src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java#L2815" class="external-link" target="_blank" rel="nofollow noopener">source</a>) runs that command just to ask Git whether sparse checkout is already enabled. Depending on how Git answers and whether you have configured "Sparse Checkout paths" in Jenkins, git-client-plugin can then run "git config core.sparsecheckout true" or "git config core.sparsecheckout false" to change the setting.</p>

<p>Also, shallow clone and sparse checkout are separate features. Shallow clone controls how much history Git clones from the remote repository to the local repository, and sparse checkout controls which files Git checks out from the local repository to the working tree.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>limors:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44598?focusedId=404506&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-404506">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Any idea how to continue on this failure in Jenkinsfile (post always does not work in this case), it seems that this plugin failure exits the build always, I wanted to notify the user automatically on this kind of failure (conflicts) in slack.</p>
</pre>
</details>

</ul>
</details>

<p>Any idea how to continue on this failure in Jenkinsfile (post always does not work in this case), it seems that this plugin failure exits the build always, I wanted to notify the user automatically on this kind of failure (conflicts) in slack.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>krisrm:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44598?focusedId=439685&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-439685">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p> Hi guys was it solved? I got similar issues:</p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/60816/60816_image-2023-07-13-08-40-45-282.png" style="border: 0px solid black" /></span>
Steps to reproduce:</p>
<ol>
	<li>Make a PR that will surely get conflicts fast</li>
	<li>Wait for some commits to come to master, that will create CONFLICTS</li>
	<li>Solve conflicts, git commit --amend --no-edit, and git push --force to PR</li>
	<li>Bitbucket will say that conflicts are solved</li>
	<li>Jenkins CI build will fail, claiming that conflicts are still there</li>
</ol>


<p>Any idea how can I fix it. Right now rebase with master doesn't work for Jenkins. It still thinks that there is a conflict :/</p>
</pre>
</details>

</ul>
</details>

<p> Hi guys was it solved? I got similar issues:</p>

<p><span class="image-wrap" style=""><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/60816/image-2023-07-13-08-40-45-282.png" style="border: 0px solid black" /></span><br/>
Steps to reproduce:</p>
<ol>
	<li>Make a PR that will surely get conflicts fast</li>
	<li>Wait for some commits to come to master, that will create CONFLICTS</li>
	<li>Solve conflicts, git commit --amend --no-edit, and git push --force to PR</li>
	<li>Bitbucket will say that conflicts are solved</li>
	<li>Jenkins CI build will fail, claiming that conflicts are still there</li>
</ol>


<p>Any idea how can I fix it. Right now rebase with master doesn't work for Jenkins. It still thinks that there is a conflict :/</p>
